### PR TITLE
Typescript configuration for Fiori elements apps

### PR DESCRIPTION
**Pull request by c-kobo**
This PR enables typescript for both apps Travel Processor and Travel Analytics.
Additionally, for the Travel Processor App, the following flexible programming model based extensions are added:
- a custom section combining a message strip and a table building block. This replaces the standard booking table section
- a controller extension: when saving, it is checked whether trees-4-tickets checkbox has been selected, a dialog is shown giving the user the option to do so before save continues.

Other changes
- Fix bad extraneous field in Travels.csv file
- Show server stdout in Karma runs
- Use `karma-ui5-transpile` to get TS working in Karma tests
- Delete obsolete `app/*/package=lock.json` files
- Unified double `karma.conf` files
- Java 21 in pipeline
- Drop Node 18 in pipeline
- Remove obsolete `odata.test.js` which also caused problems with `ts-jest`
- Remove obsolete `npm ci` calls in pipeline
- Use `deploy -out` param in Maven build to not be disturbed by extra stdout ouput from `cds-ui5-plugin`
- Replace parent pom by SpringBoot bom
- Use latest SpringBoot 3.41
- Use latest CAP Java 3.5.0

### CODE REVIEW

File: app/travel_analytics/package.json [hunk 1]
```diff
@@ -12,20 +12,29 @@
...
     "ui5-middleware-simpleproxy": "^3.2.16",
-    "ui5-task-zipper": "^3.1.4"
+    "ui5-task-zipper": "^3.1.4",
+    "@sap/ux-ui5-tooling": "1",
+    "@sapui5/types": "~1.130.0",
+    "ui5-tooling-transpile": "^3.3.7",
+    "typescript": "^5.1.6",
+    "@typescript-eslint/eslint-plugin": "^7.1.1",
+    "@typescript-eslint/parser": "^7.1.1",
+    "@sap-ux/eslint-plugin-fiori-tools": "^0.4.0"
```
Comment (devinea) [inline app/travel_analytics/package.json:38]:
```suggestion
    "@ui5/cli": "^3.0.0",
    "@sap/ux-ui5-tooling": "1",
    "@sapui5/types": "~1.130.0",
    "ui5-tooling-transpile": "^3.3.7",
    "typescript": "^5.1.6",
    "@typescript-eslint/eslint-plugin": "^7.1.1",
    "@typescript-eslint/parser": "^7.1.1",
    "@sap-ux/eslint-plugin-fiori-tools": "^0.4.0"
```

File: app/travel_analytics/package.json [hunk 2]
```diff
@@ -12,20 +12,29 @@
...
   }
```
Comment (devinea) [inline app/travel_analytics/package.json:39]:
```suggestion
  },
    "scripts": {
    "deploy-config": "npx -p @sap/ux-ui5-tooling fiori add deploy-config cf"
  }
```

File: app/travel_analytics/ui5.yaml [hunk 3]
```diff
@@ -30,3 +34,11 @@ server:
         baseUri: http://localhost:4004/analytics
         username: admin # dummy credentials for local testing
         password: admin # dummy credentials for local testing
+    - name: fiori-tools-appreload
```
Comment (devinea) [inline app/travel_analytics/ui5.yaml:37]: `ui5-middleware-simpleproxy` probably not needed?

File: app/travel_processor/package.json [hunk 4]
```diff
@@ -25,6 +28,13 @@
...
     "ui5-middleware-simpleproxy": "^3.2.16",
-    "ui5-task-zipper": "^3.1.4"
+    "ui5-task-zipper": "^3.1.4",
+    "@sap/ux-ui5-tooling": "1",
+    "@sapui5/types": "~1.130.0",
+    "ui5-tooling-transpile": "^3.3.7",
+    "typescript": "^5.1.6",
+    "@typescript-eslint/eslint-plugin": "^7.1.1",
+    "@typescript-eslint/parser": "^7.1.1",
+    "@sap-ux/eslint-plugin-fiori-tools": "^0.4.0"
```
Comment (devinea) [inline app/travel_processor/package.json:38]:
```suggestion
    "@ui5/cli": "^3.0.0",
    "@sap/ux-ui5-tooling": "1",
    "@sapui5/types": "~1.130.0",
    "ui5-tooling-transpile": "^3.3.7",
    "typescript": "^5.1.6",
    "@typescript-eslint/eslint-plugin": "^7.1.1",
    "@typescript-eslint/parser": "^7.1.1",
    "@sap-ux/eslint-plugin-fiori-tools": "^0.4.0"
```

File: app/travel_processor/package.json [hunk 5]
```diff
@@ -25,6 +28,13 @@
...
   }
```
Comment (devinea) [inline app/travel_processor/package.json:39]:
```suggestion
  },
    "scripts": {
    "deploy-config": "npx -p @sap/ux-ui5-tooling fiori add deploy-config cf"
  }
```

File: app/travel_processor/ui5.yaml [hunk 6]
```diff
@@ -30,3 +34,11 @@ server:
         baseUri: http://localhost:4004/processor
         username: admin # dummy credentials for local testing
         password: admin # dummy credentials for local testing
+    - name: fiori-tools-appreload
```
Comment (devinea) [inline app/travel_processor/ui5.yaml:37]: same. is `ui5-middleware-simpleproxy` needed anymore?

**Review by devinea — COMMENTED** [review]
@c-kobo A few small comments to align with fiori tools generation of new projects. 
I noticed a test failing here. 
http://localhost:4004/sap.fe.cap.travel/test/integration/Opa.qunit.html

File: app/travel_analytics/package.json (hunk 1 above)
Comment (Akatuoro) [reply app/travel_analytics/package.json:38]: @devinea , is the `ui5-task-zipper` included in the `@ui5/cli`? It's used in the [ui5.yaml](https://github.com/SAP-samples/cap-sflight/blob/4a0f92ba5d719074ac032afab8b33aa2d51a021b/app/travel_analytics/ui5.yaml#L17C13-L17C28)

Comment (c-kobo) [reply app/travel_analytics/package.json:38]: @devinea ui5/cli has been bumped to 4.0 by dependabot 6 months ago:
https://github.com/SAP-samples/cap-sflight/commit/edcc6164698362cf9a49164c4bb0d2cfadd9be73
Are there any issues with version 4, or is version 3 more reliable?

File: app/travel_processor/ui5.yaml (hunk 6 above)
Comment (c-kobo) [reply app/travel_processor/ui5.yaml:37]: Same here, the app won't start properly without having it in the ui5.yaml as a custom middleware.

File: app/travel_analytics/package.json (hunk 1 above)
Comment (devinea) [reply app/travel_analytics/package.json:38]: > @devinea , is the ui5-task-zipper included in the @ui5/cli? It's used in the [ui5.yaml](https://github.com/SAP-samples/cap-sflight/blob/4a0f92ba5d719074ac032afab8b33aa2d51a021b/app/travel_analytics/ui5.yaml#L17C13-L17C28)

👍 Fine to keep it if already used. npm script `deploy-config` adds it as needed. 

@c-kobo 
Based on https://sap.github.io/ui5-tooling/v4/updates/migrate-v4/ it seems it should be ok for this project. 
There are some potential issues with older ui5 versions such as 1.71 which Fiori tools needs to consider when adding @ui5/cli v4 by default.

File: app/travel_processor/package.json (hunk 4 above)
Comment (c-kobo) [reply app/travel_processor/package.json:38]: Keeping task zipper and ui5 cli 4 as discussed

File: app/travel_analytics/ui5.yaml (hunk 3 above)
Comment (devinea) [reply app/travel_analytics/ui5.yaml:37]: @c-kobo I created a PR into this branch https://github.com/SAP-samples/cap-sflight/pull/1298 where the apps load from the watch scripts at the root.

**Comment by c-kobo** [comment]
Wrt to the failing OPA tests, my colleague @ARiesterer will try to take a look into it.

**Comment by c-kobo** [comment]
> Wrt to the failing OPA tests, my colleague @ARiesterer will try to take a look into it.

OPA tests should be fine again now.

**Comment by c-kobo** [comment]
Jest tests still failing.

**Comment by c-kobo** [comment]
@stewsk @chgeo the jest tests fail due to an incompability with cds-plugin-ui5.
Seems to be an issue with jest and node in combination with plugins.
@petermuessig described a workaround here (didn't work for me though):
https://github.com/ui5-community/ui5-ecosystem-showcase/issues/901#issuecomment-1785216920

**Review by chgeo — APPROVED** [review]
Let's merge this as this is to be shown tomorrow in CAP product day.

Aligned with @Akatuoro that he rebases his TS simplificiations on top of this one.

**Comment by beckermarc** [comment]
@chgeo Is there any reason, why this PR removed the Spring Boot parent POM from the Java build? This actually broke one of our scenarios which requires Maven profiles that come with the Spring Boot parent POM. 

Sadly this went undetected until now, even though we thought we tested this scenario nightly, but the Maven build just did less than expected, because the said Maven profiles were removed, and still was successful.

**Comment by chgeo** [comment]
@beckermarc sorry, I have no idea other than the [commit message](https://github.com/SAP-samples/cap-sflight/pull/1286/commits/5be014613bfa49d3b3cdb80cf3134062160bffaf) which indicates there was a build issue.  
Maybe I was misled there, but certainly followed the parent-less structure that the current archetype generates.  This means we want to deviate here?

**Comment by beckermarc** [comment]
> This means we want to deviate here?

Yep, because we use the GraalVM Native Image feature in this project, which requires the Spring Boot parent pom. We don't enable projects for that feature by default with `cds init`. 

I will look into a PR to restore the previous behaviour next week. No worries :)

**Comment by chgeo** [comment]
> I will look into a PR to restore the previous behaviour next week. No worries :)

Thanks!

**Pull request merged**